src/core/bpf-lsm.c

```diff
@@ -0,0 +1,375 @@
+        r = bpf_object_new(restrict_fs_hexdump_buffer, sizeof(restrict_fs_hexdump_buffer), &obj);
+        if (r < 0) {
+                log_error_errno(r, "Failed to create BPF object from hexdump buffer: %m");
+                return supported = 0;
+        }
+
+        /* Dummy map to satisfy the verifier */
+        inner_map_fd = bpf_create_map(BPF_MAP_TYPE_HASH, sizeof(uint32_t), sizeof(uint32_t), 128, 0);
+        if (inner_map_fd < 0) {
+                log_debug_errno(errno, "Failed to create BPF map");
+                return supported = 0;
+        }
+
+        r = bpf_object_set_inner_map_fd(obj, "cgroup_hash", inner_map_fd);
+        if (r < 0) {
+                log_debug_errno(r, "Failed to set inner map fd: %m");
+                return supported = 0;
+        }
+
+        r = bpf_object_load(obj);
+        if (r < 0) {
+                log_debug_errno(errno, "Failed to load LSM BPF program: %m");
+                return supported = 0;
+        }
+
+        return supported = 1;
+}
+
+int lsm_bpf_setup(void) {
+        _cleanup_(bpf_object_freep) struct bpf_object *obj = NULL;
+        struct bpf_program *prog = NULL;
+        struct bpf_link *link = NULL;
+        _cleanup_close_ int inner_map_fd = -1;
+        int outer_map_fd;
+        int r;
+
+        r = bpf_object_new(restrict_fs_hexdump_buffer, sizeof(restrict_fs_hexdump_buffer), &obj);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create BPF object from hexdump buffer: %m");
+
+        /* Dummy map to satisfy the verifier */
+        inner_map_fd = bpf_create_map(BPF_MAP_TYPE_HASH, sizeof(uint32_t), sizeof(uint32_t), 128, 0);
+        if (inner_map_fd < 0)
+                return log_error_errno(errno, "Failed to create BPF map: %m");
+
+        r = bpf_object_set_inner_map_fd(obj, "cgroup_hash", inner_map_fd);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set inner map fd: %m");
+
+        r = bpf_object_load(obj);
+        if (r < 0)
+                return log_error_errno(errno, "Failed to load LSM BPF program: %m");
+
+        /* Close dummy map */
+        inner_map_fd = safe_close(inner_map_fd);
+
+        r = bpf_object_find_program_by_title(obj, LSM_PROG_SECTION_NAME, &prog);
+        if (r < 0)
+                return log_error_errno(r, "Failed to find LSM BPF program '%s': %m", LSM_PROG_SECTION_NAME);
+
+        link = bpf_program__attach_lsm(prog);
+        if (!link)
+                return log_error_errno(errno, "Failed to attach BPF LSM program: %m");
+
+        log_info("LSM BPF program attached");
+
+        outer_map_fd = bpf_object_get_map_fd(obj, "cgroup_hash");
+        if (outer_map_fd < 0)
+                return log_error_errno(outer_map_fd, "Failed to get map fd: %m");
+
+        r = mkdir_parents(LSM_MAP_BPFFS_PATH, 0700);
+        if (r < 0)
+                return log_error_errno(r, "Failed create systemd bpffs directory: %m");
+
+        r = bpf_obj_pin(outer_map_fd, LSM_MAP_BPFFS_PATH);
```

Comment:
Pinning step can be dropped altogether if `outer_map_fd` is stored. Alternatively bpf skeleton generated object provide `bpf_map` pointer. Similar to `bpf_link`, `Manager` looks like a proper place to store either object or fd.
IMO using bpffs questionable in this context, since it's a security feature and with bpffs too much container internals are shared with host. Also bpffs may not be mounted.

Reply:
Make sense, I'll give it a go and change thus to use the BPF skeleton approach.